Mathlib/Analysis/Calculus/FDeriv/Norm.lean

```diff
@@ -0,0 +1,200 @@
+  · exact differentiableAt_norm_smul ht |>.2 h
+
+theorem DifferentiableAt.fderiv_norm_self {x : E} (h : DifferentiableAt ℝ (‖·‖) x) :
+    fderiv ℝ (‖·‖) x x = ‖x‖ := by
+  rw [← h.lineDeriv_eq_fderiv, lineDeriv]
+  have this (t : ℝ) : ‖x + t • x‖ = |1 + t| * ‖x‖ := by
+    rw [← norm_eq_abs, ← norm_smul, add_smul, one_smul]
+  simp_rw [this]
+  rw [deriv_mul_const]
+  · conv_lhs => enter [1, 1]; change _root_.abs ∘ (fun t ↦ 1 + t)
+    rw [deriv.comp, deriv_abs, deriv_const_add]
+    · simp
+    · exact differentiableAt_abs (by norm_num)
+    · exact differentiableAt_id.const_add _
+  · exact (differentiableAt_abs (by norm_num)).comp _ (differentiableAt_id.const_add _)
+
+variable (x t) in
+theorem fderiv_norm_smul :
+    fderiv ℝ (‖·‖) (t • x) = (SignType.sign t : ℝ) • (fderiv ℝ (‖·‖) x) := by
+  by_cases hE : Nontrivial E
+  · by_cases hd : DifferentiableAt ℝ (‖·‖) x
+    · obtain rfl | ht := eq_or_ne t 0
+      · simp only [zero_smul, _root_.sign_zero, SignType.coe_zero]
+        exact fderiv_zero_of_not_differentiableAt <| not_differentiableAt_norm_zero E
+      · rw [(hd.hasFDerivAt.hasFDerivAt_norm_smul ht).fderiv]
+    · rw [fderiv_zero_of_not_differentiableAt hd, fderiv_zero_of_not_differentiableAt]
+      · simp
+      · exact mt DifferentiableAt.differentiableAt_norm_of_smul hd
+  · rw [not_nontrivial_iff_subsingleton] at hE
+    simp_rw [(hasFDerivAt_of_subsingleton _ _).fderiv, smul_zero]
+
+theorem fderiv_norm_smul_pos (ht : 0 < t) :
+    fderiv ℝ (‖·‖) (t • x) = fderiv ℝ (‖·‖) x := by
+  simp [fderiv_norm_smul, ht]
+
+theorem fderiv_norm_smul_neg (ht : t < 0) :
+    fderiv ℝ (‖·‖) (t • x) = -fderiv ℝ (‖·‖) x := by
+  simp [fderiv_norm_smul, ht]
+
+theorem norm_fderiv_norm [Nontrivial E] (h : DifferentiableAt ℝ (‖·‖) x) :
+    ‖fderiv ℝ (‖·‖) x‖ = 1 := by
+  have : x ≠ 0 := fun hx ↦ not_differentiableAt_norm_zero E (hx ▸ h)
+  refine le_antisymm (NNReal.coe_one ▸ norm_fderiv_le_of_lipschitz ℝ lipschitzWith_one_norm) ?_
+  apply le_of_mul_le_mul_right _ (norm_pos_iff.2 this)
+  calc
+    1 * ‖x‖ = fderiv ℝ (‖·‖) x x := by rw [one_mul, h.fderiv_norm_self]
+    _ ≤ ‖fderiv ℝ (‖·‖) x x‖ := le_norm_self _
+    _ ≤ ‖fderiv ℝ (‖·‖) x‖ * ‖x‖ := le_opNorm _ _
+
+/-- In a real finite-dimensional normed vector space,
+  the norm is almost everywhere differentiable. -/
+theorem ae_differentiableAt_norm [FiniteDimensional ℝ E] [MeasurableSpace E] [BorelSpace E]
+    {μ : MeasureTheory.Measure E} [μ.IsAddHaarMeasure] :
+  ∀ᵐ x ∂μ, DifferentiableAt ℝ (‖·‖) x := lipschitzWith_one_norm.ae_differentiableAt
+
+/-- In a real finite-dimensional normed vector space,
+  the set of points where the norm is differentiable at is dense. -/
+theorem dense_differentiableAt_norm [FiniteDimensional ℝ E] :
+    Dense {x : E | DifferentiableAt ℝ (‖·‖) x} :=
+  let _ : MeasurableSpace E := borel E
+  have _ : BorelSpace E := ⟨rfl⟩
+  let w := Basis.ofVectorSpace ℝ E
+  MeasureTheory.Measure.dense_of_ae (ae_differentiableAt_norm (μ := w.addHaar))
```

Comment:
Just because of these two lemmas, you have heavy imports in your file (measure theory, Rademacher). To keep more reasonable imports, what do you think of moving these to the Rademacher files, as illustrations of the Rademacher theorem in a concrete case?